src/channel-qvm.lisp

```diff
@@ -0,0 +1,204 @@
+;;;; channel-qvm.lisp
+;;;;
+;;;; Authors: Sophia Ponte, Nikolas Tezak, Erik Davis
+
+(in-package #:qvm)
+
+;;; Implements a QVM that can interpret a rule-based noise model and
+;;; apply kraus operators to it's state.
+
+(defgeneric apply-all-kraus-maps (qvm instr kraus-maps)
+  (:documentation "Applies every kraus-map in the list KRAUS-MAPS to the state of the system."))
+
+
+(defgeneric apply-kraus-map (qvm instr kraus-ops)
+  (:documentation "Applies noise from a kraus map to the system. Randomly select a kraus operator from the kraus map KRAUS-OPS to apply to the state of the system."))
+
+
+;;; The CHANNEL-QVM requires an explicitly defined NOISE-MODEL. This
+;;; NOISE-MODEL should be a complete specification of the different
+;;; types of noise that should be applied throughout the execution of
+;;; a QUIL program, as well as when in the program's execution the
+;;; noise should be applied. These decisions are made during
+;;; TRANSITION. At each new instruction, the CHANNEL-QVM checks to see
+;;; if any NOISE-RULES are matched by the instuction, and if so, the
+;;; CHANNEL-QVM applied the corresponding noise.
+(defclass channel-qvm (pure-state-qvm)
+  ((original-amplitude-pointer
+    :reader original-amplitude-pointer)
+   (trial-amplitudes
+    :accessor %trial-amplitudes)
+   (noise-model
+    :initarg :noise-model
+    :accessor noise-model
+    :initform nil))
+  (:documentation "The CHANNEL-QVM is a QVM that supports a fully explicit NOISE-MODEL. The NOISE-MODEL is a list of NOISE-RULES, where each NOISE-RULE describes a type of noise and at which points in a program's execution that type of noise should be applied."))
+
+
+(defmethod initialize-instance :after ((qvm channel-qvm) &rest args)
+  ;; Initializes an instance of a CHANNEL-QVM"
+  (declare (ignore args))
+  (setf
+   ;; Initialize the trial-amplitudes to an empty array of the correct
+   ;; size
+   (%trial-amplitudes qvm) (make-lisp-cflonum-vector (expt 2 (number-of-qubits qvm)))
+   ;; Save a pointer to the originally provided memory
+   (slot-value qvm 'original-amplitude-pointer) (amplitudes qvm)))
+
+
+(defmethod run :after ((qvm channel-qvm))
+  ;; Only copy if we really need to.
+  (when (requires-swapping-p qvm)
+    ;; Copy the correct amplitudes into place.
+    (copy-wavefunction (amplitudes qvm) (original-amplitude-pointer qvm))
+    ;; Get the pointer back in home position. We want to swap them,
+    ;; not overwrite, because we want the scratch memory to be intact.
+    (rotatef (amplitudes qvm) (%trial-amplitudes qvm))))
+
+
+(defun requires-swapping-p (qvm)
+  "Does the  QVM require swapping of internal pointers?"
+  (and (not (eq (amplitudes qvm) (original-amplitude-pointer qvm)))
+       #+sbcl (eq ':foreign (sb-introspect:allocation-information
+                             (original-amplitude-pointer qvm)))))
+
+
+(defmethod transition ((qvm channel-qvm) (instr quil:gate-application))
```

Comment:
As with `transition` in `basic-noise-qvm`, I wonder if this is the best method combination. Consider what this would look like using `:around`.

Reply:
https://github.com/rigetti/qvm/pull/195#discussion_r342281033